### ISSUE TITLE
DSP: MVE arm_biquad_cascade_df1_f16 out of bounds access to destination buffer

### ISSUE THREAD

**Issue by stephanosio** [ARM-software/CMSIS_5#1475]
The [MVE `arm_biquad_cascade_df1_f16` implementation](https://github.com/ARM-software/CMSIS_5/blob/347ca08b52808350c6328cd6a28f42f75dcddaad/CMSIS/DSP/Source/FilteringFunctions/arm_biquad_cascade_df1_f16.c#L54) has two major issues:

1. It requires the destination buffer to be initialised to the value of 0 -- this in itself is not a problem, but noting that the scalar implementation does not require this, this behavioural difference should be noted in the documentation.
2. It attempts to access data beyond the destination buffer -- for instance, when `blockSize` is 100, it requires the `pDst[0:99]` as well as the `pDst[100:101]` to be initialised to 0. If the `pDst[100]` and `pDst[101]` are not zero, the `pDst[98]` and `pDst[99]` will contain invalid results.

This can be reproduced by setting `output[100]`  and `output[101]` to a non-zero value here:

https://github.com/ARM-software/CMSIS_5/blob/3fa74ec41ff7b6cbbb941dd9c6ba1048a2606536/CMSIS/DSP/Testing/Source/Tests/BIQUADF16.cpp#L26

UPDATE: The same issue can be observed for `arm_biquad_cascade_df1_f32`.

**Comment by christophe0606** [comment]
@stephanosio It should not require the destination buffer to be set to 0 nor it should write after the end of buffer.

I'll look at it because it is definitely not right. By quickly reviewing the source code I don't see any place where dst is read so I don't see why it should need to be initialized to zero. So, I will need more time to identify the issue.

Which version of CMSIS-DSP ?

Have you checked if it is not another problem like some buffer overlap or memory corruption in the test ?

The MVE version must be initialized differently. Documentation is saying:

> For MVE code, an additional buffer of modified coefficients is required.
> Its size is numStages and each element of this buffer has type arm_biquad_mod_coef_f16. So, its total size is 96*numStages float16_t elements.
> 

And there is the new `arm_biquad_cascade_df1_mve_init_f16` to use instead.

But if you use CMSIS-DSP tests then it should already be the case. I'll have to check also the test part that may be wrong. (And it is not detecting the write after end of buffer which is normally part of the tests).

**Comment by stephanosio** [comment]
> Which version of CMSIS-DSP ?

1.9.0 (CMSIS 5.8.0), but should be the same for 1.10.0 since there were no functional changes since 1.9.0.

> And there is the new arm_biquad_cascade_df1_mve_init_f16 to use instead.

Yes, I am using it for MVE testing: https://github.com/zephyrproject-rtos/zephyr/blob/d1eea105378e98874190a81c8415a80414742997/tests/lib/cmsis_dsp/filtering/src/biquad_f16.c#L51

> Have you checked if it is not another problem like some buffer overlap or memory corruption in the test ?

I doubt it, but I will double check also.

FYI, here, without `length + 2` (note that `calloc` allocates and **initialises** to 0), the `output_buf[98]` and `output_buf[99]` will be NaN for some reason.
https://github.com/zephyrproject-rtos/zephyr/blob/d1eea105378e98874190a81c8415a80414742997/tests/lib/cmsis_dsp/filtering/src/biquad_f16.c#L43-L45

**Comment by stephanosio** [comment]
@christophe0606 This looks very suspicious:

https://github.com/ARM-software/CMSIS_5/blob/3fa74ec41ff7b6cbbb941dd9c6ba1048a2606536/CMSIS/DSP/Source/FilteringFunctions/arm_biquad_cascade_df1_f16.c#L152-L170

With the `blockSize` of 50 (I incorrectly said 100 in the initial comment; the total input length is 100), the remainder should be 2 (50 mod 8); yet, it reads a total of 8 elements/6 extra elements into `X0-X7` from `pIn`, of which `X2-X7` are beyond the input buffer and will contain invalid values. Then, it goes on to MAC `X0-X7` into the `accVec`, which is then stored to the output buffer.

Note that from second stage and on, the `pIn` is set to `pDst`, so this effectively leads to the function accessing beyond the end of the output buffer.

p.s. I have not actually debugged it line by line. This was all done in my head, so the analysis might not be completely right.

**Comment by christophe0606** [comment]
@stephanosio Thanks for the analysis. I am going to look at it.

**Comment by christophe0606** [comment]
@stephanosio So far I have not managed to reproduce the issue.
I am corrupting the dst in the test with:

```
memset(outp,1,(blockSize+2)*sizeof(float16_t));
arm_biquad_cascade_df1_f16(&this->Sdf1,inputp,outp,blockSize);
```
 It is not changing anything to the result of the computation.

I have looked at the suspicious part and in fact the magic is that the pCoeffs has following structure:
```
0 0 ... 0   0  A1
0 0 ... 0   B1 A2
0 0 ... C1  B2 A3
```

Then the vfmaq are done starting from X7 to X0.
So the first multiply by X7 is only impacting lane 7
X6 is only impacting the result starting from lane 6 ...

...

and X0 is only impacting the result from lane 0

After we have a switch case which is only reading (in your case) the lane 0 and 1.

So the result is indeed impacted only by 2 values and not by the values read outside of the buffer.

I have confirmed that the *pIn++ are indeed reading the 2 first values from the buffer.

So, the documentation must be updated to explain that some reads are done after the end of the dst buffer.

But from a computation point of view it looks ok.

Perhaps you have problems with `arm_biquad_cascade_df1_mve_init_f16` and initialization of the coefficients ?

Note also that f16 computations are more likely to saturate. So depending on the coefficients and the blocksize having a NaN in the result does not always means there is a problem in the algorithm.
